root/usr/share/ruby/vendor_ruby/discovery.rb

```diff
@@ -231,21 +231,52 @@ def get_mac(interface = 'primary')
 
 def get_ipv4(interface = 'primary')
   wait = cmdline('fdi.nmwait', 120)
-  `nmcli -w #{wait} -t -f IP4.ADDRESS con show #{interface} 2>/dev/null`.split(":", 2)[1].chomp rescue "N/A"
+
+  begin
+    `nmcli -w #{wait} -t -f IP4.ADDRESS con show #{interface} 2>/dev/null`.split(':', 2)[1].chomp
+  rescue StandardError
+    'N/A'
+  end
 end
 
 def get_ipv6(interface = 'primary')
   wait = cmdline('fdi.nmwait', 120)
-  `nmcli -w #{wait} -t -f IP6.ADDRESS con show #{interface} 2>/dev/null`.split(":", 2)[1].chomp rescue "N/A"
+
+  begin
+    `nmcli -w #{wait} -t -f IP6.ADDRESS con show #{interface} 2>/dev/null`.split(':', 2)[1].chomp
+  rescue StandardError
+    'N/A'
+  end
 end
 
 def detect_ipv4_credentials(interface)
   res = {}
-  str = `nmcli -t -f IP4.ADDRESS,IP4.GATEWAY,IP4.DNS con show #{interface} 2>/dev/null`
-  return ["", "", ""] if $? != 0
-  str.each_line { |x| kv = x.split(':'); res[kv[0]] = kv[1].chomp }
-  [res["IP4.ADDRESS[1]"] || '', res["IP4.GATEWAY"] || '', res["IP4.DNS[1]"] || '']
-rescue => e
+  str = `nmcli -t -f IP4.ADDRESS,IP4.GATEWAY,IP4.DNS con show '#{interface}' 2>/dev/null`
+  return [nil, nil, nil] if $? != 0
+
+  str.each_line do |x|
+    kv = x.split(':')
+    res[kv[0]] = kv[1].chomp
+  end
+
+  [res['IP4.ADDRESS[1]'], res['IP4.GATEWAY'], res['IP4.DNS[1]']]
+rescue StandardError => e
+  log_err e.message
+  [nil, nil, nil]
+end
+
+def detect_ipv6_credentials(interface)
+  res = {}
+  str = `nmcli -t -f IP6.ADDRESS,IP6.GATEWAY,IP6.DNS con show '#{interface}' 2>/dev/null`
+  return [nil, nil, nil] if $? != 0
+
+  str.each_line do |x|
+    kv = x.split(':')
+    res[kv[0]] = kv[1..].join(':').chomp
+  end
+
+  [res['IP6.ADDRESS[1]'], res['IP6.GATEWAY'], res['IP6.DNS[1]']]
```

Comment:
I think we can filter out the link-local IPs here. Instead of always taking `[1]`, we can iterate through them and pick the first that is not link-local

Reply:
Not sure how to do that.

RIght now the check is in the `if cidr6 && !cidr6.link_local?`

Reply:
The result that you get out of command is something like this:
```
IP6.ADDRESS[1]:                         2620:1:1:1::1/64
IP6.ADDRESS[2]:                         fe80::1:2:3:4/64
```

You should iterate through all the lines and select only the one that is not link local instead of always taking the `IP6.ADDRESS[1]` field

Reply:
oh yeah now I get it. Ok will do.